File: choeunwoo/week6/boj_11660.c
#include <stdio.h>

int N, M;
int arr[1025][1025];  // 입력 배열
int psum[1025][1025]; // 2D 누적합

int main() {
    scanf("%d %d", &N, &M); // N, M 입력

    // 배열 값 입력
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= N; j++) {
            scanf("%d", &arr[i][j]);
        }
    }

    // 2D 누적합 계d
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= N; j++) {
            psum[i][j] = psum[i - 1][j]
                       + psum[i][j - 1]
                       - psum[i - 1][j - 1]
                       + arr[i][j];
        }
    }

    // 질의 처리
    for (int q = 0; q < M; q++) {
        int x1, y1, x2, y2;
        scanf("%d %d %d %d", &x1, &y1, &x2, &y2);
        int result = psum[x2][y2]
                   - psum[x1 - 1][y2]
                   - psum[x2][y1 - 1]
                   + psum[x1 - 1][y1 - 1];
        printf("%d\n", result);
    }

    return 0;
}

File: choitaejin/week6/boj_11660.js
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
});

let n, m;
let lineCnt = 0;
let arr = [];
let arrIdx = 0;
let cmd = [];
let cmdIdx = 0;
let prefix = [];

rl.on('line', (line) => {
    if (lineCnt == 0) {
        const temp = line.split(' ').map(Number);
        n = temp[0];
        m = temp[1];

        arr = Array(n).fill().map(() => Array(n).fill(0));
        cmd = Array(m).fill().map(() => Array(4).fill(0));
        prefix = Array(n+1).fill().map(() => Array(n+1).fill(0));

        lineCnt++;
    }
    else if (lineCnt > 0 && lineCnt < n+1) {
        arr[arrIdx] = line.split(' ').map(Number);
        arrIdx++;
        lineCnt++;
    }
    else if (lineCnt > n && lineCnt <= n + m) {
        cmd[cmdIdx] = line.split(' ').map(Number);
        cmdIdx++;
        lineCnt++;
    }
    else rl.close();
});
rl.on('close', () => {
    for (let i = 1; i <=n; i++) {
        for (let j = 1; j <= n; j++) {
            prefix[i][j] = prefix[i-1][j] + prefix[i][j-1] - prefix[i-1][j-1] + arr[i-1][j-1];
        }
    }

    for (let k = 0; k < m; k++) {
        let x1 = cmd[k][0];
        let y1 = cmd[k][1];
        let x2 = cmd[k][2];
        let y2 = cmd[k][3];

        const result = prefix[x2][y2]           
                      - prefix[x1-1][y2]         
                      - prefix[x2][y1-1]         
                      + prefix[x1-1][y1-1];     

       console.log(result);
    }

    process.exit();
});